Package requirements: don't apply to build-only or external packages

scheibelp opened the pull request:
You can now augment a requirement to designate that it should not apply for build-only or external packages:

```
packages:
  all:
    require:
    - spec: "@1.0"
      turn_off_for: ["externals", "build"]
```

(example taken from unit test added here)

it is assumed this is useful because for example you may want to force building your top-level packages with a specific compiler, but don't care what CMake (or it's dependencies) build with. Likewise, we may want to accurately represent externals that were built with `%gcc`, but force everything we're building to build with `%intel`.

### Old: prior to Jan. 19 (the schema options changed):

```
packages:
  all:
    require:
    - spec: "@1.0"
      root_only: true
      applies_for_externals: false
```

scheibelp committed:
initial (failing) test

scheibelp committed:
intermediate work, test passing (with hacks)

scheibelp committed:
was testing nodeness wrong; testing membership of 'build' unification set seems more loose than checking for build-only

scheibelp committed:
style auto fix + flake import complaint

scheibelp committed:
test skipping application of requirements to externals

scheibelp committed:
TODOs

scheibelp committed:
track set of link/run deps that are traceable to root

scheibelp committed:
check link deps of build-only deps

scheibelp committed:
minor cleanup

tgamblin commented on lib/spack/spack/schema/packages.py:
```diff
@@ -41,6 +41,8 @@
                             "spec": {"type": "string"},
                             "message": {"type": "string"},
                             "when": {"type": "string"},
+                            "root_only": {"type": "boolean"},
+                            "applies_for_externals": {"type": "boolean"},
```
I think this should be a bit more flexible / extensible and we should use a list of strings rather than boolean options, e.g.:

```yaml
packages:
  all:
    require:
    - spec: "@1.0"
      include:: ["externals", "roots"]
```

I'm not sure these are the only elements we'd want but it's a start.  The default would be to include everything.

tgamblin commented on lib/spack/spack/schema/packages.py:
```diff
@@ -41,6 +41,8 @@
                             "spec": {"type": "string"},
                             "message": {"type": "string"},
                             "when": {"type": "string"},
+                            "root_only": {"type": "boolean"},
```
I don't know that `roots` is the right name here. The PR says this is for build-only packages.  What is the relation here between those and roots?

scheibelp committed:
merge from develop (seemingly minor conflicts)

scheibelp committed:
reorg two boolean flags into a list of enums; rename flags for clarity

scheibelp committed:
auto style fix

scheibelp committed:
rename asp function for clarity

scheibelp commented:
@tgamblin I've updated this according to https://github.com/spack/spack/pull/41406#discussion_r1457087737 (although the semantics I implemented are a blacklist vs. a whitelist): for now I've replaced the multiple new schema properties with a single list of enums; still restrictive for now but easier to expand.

Also I've replaced the confusingly named `root_only` with `turn_off_for: [build]`

scheibelp committed:
merge from develop (somewhat involved conflicts based on relocation, but semantically not a major issue)

scheibelp committed:
rule instantiations require explicit parameters

scheibelp committed:
allow default values for namedtuple

scheibelp committed:
Revert "allow default values for namedtuple"

This reverts commit 6315d210e4bd531ace08eafdc5d87f72cfe51d7d.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -55,6 +55,7 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
+% What does this mean?
```
This means that a "build" dependency moves across unification sets (from `SetID` of the parent node to `("build", node(X, Child))` of the child node).
```suggestion
```

scheibelp committed:
Merge branch 'develop' into features/selective-requirements

scheibelp committed:
remove duplicate derivation that is available w/ root unification set

scheibelp committed:
merge from develop (minor conflicts)
